pinot-segment-local/src/main/java/org/apache/pinot/segment/local/indexsegment/mutable/MutableSegmentImpl.java

```diff
@@ -1467,7 +1467,8 @@ private class IndexContainer implements Closeable {
 
     DataSource toDataSource() {
       return new MutableDataSource(_fieldSpec, _numDocsIndexed, _numValuesInfo._numValues,
-          _numValuesInfo._maxNumValuesPerMVEntry, _partitionFunction, _partitions, _minValue, _maxValue, _forwardIndex,
+          _numValuesInfo._maxNumValuesPerMVEntry, _statsHistory.getEstimatedCardinality(_fieldSpec.getName()),
```

Comment:
I wouldn't read the cardinality from the `_statsHistory` for the following reasons:
1. It can be expensive, and calling it for every column at query time can cause too much overhead
2. It won't reflect the cardinality for the current consumed records

Consider using the `_dictionary.length()` when dictionary is available, `-1` if not

Reply:
Done